### PR TITLE
[RemoteDev] Support remote development

### PR DESCRIPTION
**Pull request by TongboZhang**
Remote Development can allow users to develop locally with PAI's resources and improve users’ development experience.

Basicaly, the main process is to ask PAI for a container with resources, and use local editors or IDEs to develope and debug in this container using ssh.

Support IDEs or editors:
- CLI
- Visual Studio Code
- PyCharm

Support OS:
- Ubuntu 16.04
- Ubuntu 18.04
- Windows 10

### CODE REVIEW

File: contrib/remote-dev/README.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+│   └── job.template             # PAI template job
+└── start.py                     # start script
+```
+
+## 3. Dependencies
+
+- This sub project depend on [PAI's Python SDK](https://github.com/microsoft/pai/tree/master/contrib/python-sdk), please install and test the sdk first.
+
+- The `start.py` script requires python3, and we only tested it on `py3.5+` environment, and ruquired following moduls:
+    - requests
```
Comment (Binyang2014) [inline contrib/remote-dev/README.md]: write down the `pip install` command?

File: contrib/remote-dev/README.md [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+      gpu: 0
+      cpu: 1
+      memoryMB: 2048
+    commands:
+      - '`#REMOTE_DEV_START`'
+      - apt-get update
+      - apt-get install -y nfs-common cifs-utils sshpass wget
+      - umask 000
+      - mkdir -p /root/workspace
+      - mount.cifs //sharepath /root/workspace -o username=***,password=***
```
Comment (Binyang2014) [inline contrib/remote-dev/README.md]: Do we need to change set the value for username & password. Or we just leave it as `***`?

File: contrib/remote-dev/README.md [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+### 3.3 Configure Editors and IDEs
+
+#### 3.3.1 CLI
+
+For CLI usage, just ssh into your container and you can do anything you want. Your workspace will be mounted at `/root/workspace`.
+
+#### 3.3.2 PyCharm Professional
+
+For PyCharm Professional usage, you can configure PyCharm `Deployment` and `Project Interpreter` according to the following animation.
```
Comment (Binyang2014) [inline contrib/remote-dev/README.md]: Please clarify that it didn't work for community version

File: contrib/remote-dev/README.md [hunk 4]
```diff
@@ -0,0 +1,165 @@
+# Remote Development
+
+## 1. Overview
+
+Remote Development can allow users to develop locally with PAI's resources and improve users’ development experience.
+
+Basicaly, the main process is to ask PAI for a container with resources, and use local editors or IDEs to develope and debug in this container using ssh.
```
Comment (Binyang2014) [inline contrib/remote-dev/README.md]: Make a list for the main process?

File: contrib/remote-dev/README.md [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+    - configparser
+    - subprocess
+
+## 3. Usage
+
+This section will give guidance about usage.
+
+### 3.1 Preparation
+
+First, please configure the vars in `.env.template` and rename it to `.env`.
```
Comment (Binyang2014) [inline contrib/remote-dev/README.md]: Can give a mock example for the config?

File: contrib/remote-dev/README.md (hunk 1 above)
Comment (TongboZhang) [reply contrib/remote-dev/README.md]: These modules are installed by default. I will remove this module request.

File: contrib/remote-dev/README.md (hunk 2 above)
Comment (TongboZhang) [reply contrib/remote-dev/README.md]: It should be changed. I will make this more clear.

File: contrib/remote-dev/README.md (hunk 3 above)
Comment (TongboZhang) [reply contrib/remote-dev/README.md]: This can only work in professional version. Will make it more clear.

File: contrib/remote-dev/README.md (hunk 4 above)
Comment (TongboZhang) [reply contrib/remote-dev/README.md]: OK

File: contrib/remote-dev/README.md (hunk 5 above)
Comment (TongboZhang) [reply contrib/remote-dev/README.md]: OK

File: contrib/remote-dev/README.md [hunk 6]
```diff
@@ -0,0 +1,183 @@
+# Remote Development
+
+## 1. Overview
+
+`Remote Development` can allow users to develop locally with PAI's resources and improve users’ development experience.
```
Comment (abuccts) [inline contrib/remote-dev/README.md]: remove `can`
`’` -> `'`

File: contrib/remote-dev/README.md [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+
+The main process of `Remote Development`:
+- Ask PAI for a container with resources using `start.py`
+- Get SSH info
+- Configure local editors or IDEs
+- Develop and debug in this container like local
+
+Support IDEs or editors:
```
Comment (abuccts) [inline contrib/remote-dev/README.md]: `Support` -> `Supported`

File: contrib/remote-dev/README.md [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+- CLI
+- Visual Studio Code
+- PyCharm
+
+Support OS:
```
Comment (abuccts) [inline contrib/remote-dev/README.md]: `Support` -> `Supported`

File: contrib/remote-dev/README.md [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+
+For more details, please refer to [VS code Remote Development](https://www.jetbrains.com/help/pycharm/remote-debugging-with-product.html).
+
+## 4. Advance
+
+You can also deploy an NFS/SMB server on your host, and mount it in PAI's container by changing the `commands` of the job template.
+
+In ubuntu, you can run a container with NFS4 server, and use `mount -t nfs4 <hostip>:<sharepath> <mountpoint>` to mount. And in windows, you can use built-in SMB, and use `mount -t cifs //<hostip>/<sharename> <mountpoint> -o vers=3.0,username=<username>,password=<password>,domain=<domain>` to mount.
+
+## License
```
Comment (abuccts) [inline contrib/remote-dev/README.md]: remove

File: contrib/remote-dev/conf/job.template [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    dockerImage: docker_image_0
+    resourcePerInstance:
+      gpu: 0
+      cpu: 1
+      memoryMB: 2048
+    commands:
+      - '`#REMOTE_DEV_START`'
+      - '`#Auto generated code, please do not modify`'
+      - apt-get update
+      - apt-get install -y nfs-common cifs-utils sshpass wget
```
Comment (abuccts) [inline contrib/remote-dev/conf/job.template]: do you need `sshpass`?

File: contrib/remote-dev/conf/job.template [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+    completion:
+      minFailedInstances: 1
+      minSucceededInstances: 1
+    dockerImage: docker_image_0
+    resourcePerInstance:
+      gpu: 0
+      cpu: 1
+      memoryMB: 2048
+    commands:
+      - '`#REMOTE_DEV_START`'
```
Comment (abuccts) [inline contrib/remote-dev/conf/job.template]:
```suggestion
      # REMOTE_DEV_START
```

File: contrib/remote-dev/conf/job.template [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+      - '`#Auto generated code, please do not modify`'
```
Comment (abuccts) [inline contrib/remote-dev/conf/job.template]:
```suggestion
      # Auto generated code, please do not modify
```

File: contrib/remote-dev/conf/job.template [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+      - apt-get update
+      - apt-get install -y nfs-common cifs-utils sshpass wget
+      - umask 000
+      - mkdir -p /root/workspace
+      - mount.cifs //sharepath /root/workspace -o username=***,password=***
+      - sleep 18000
+      - '`#REMOTE_DEV_END`'
```
Comment (abuccts) [inline contrib/remote-dev/conf/job.template]:
```suggestion
      # REMOTE_DEV_END
```

File: contrib/remote-dev/conf/job.template [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+      cpu: 1
+      memoryMB: 2048
+    commands:
+      - '`#REMOTE_DEV_START`'
+      - '`#Auto generated code, please do not modify`'
+      - apt-get update
+      - apt-get install -y nfs-common cifs-utils sshpass wget
+      - umask 000
+      - mkdir -p /root/workspace
+      - mount.cifs //sharepath /root/workspace -o username=***,password=***
```
Comment (abuccts) [inline contrib/remote-dev/conf/job.template]: please use `secrets` field for username and password

File: contrib/remote-dev/start.py [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+def read_inpur():
+  while (True):
+    answer = input("Do you wish to continue (y/n)?")
+    if answer == "Y" or answer == "y":
+      break
+    elif answer == "N" or answer == 'n':
+      exit(0)
+
+
+def print_output(info):
```
Comment (abuccts) [inline contrib/remote-dev/start.py]: please change to `logger`

Comment (abuccts) [reply contrib/remote-dev/start.py]: reference: https://docs.python.org/3/library/logging.html

File: contrib/remote-dev/start.py [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+  f2 = open(userdir + conf_file + ".yaml", 'w+')
+  for line in f1.readlines():
+    f2.write(replace(line))
+  f1.close()
+  f2.close()
+
+# display PAI cluster
+print_output("Display PAI cluster")
+subprocess.call("opai cluster list", shell=is_shell)
+read_inpur()
```
Comment (abuccts) [inline contrib/remote-dev/start.py]: `read_input()`

File: contrib/remote-dev/start.py [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+import time
+import random
+import string
+import requests
+import configparser
+import subprocess
+import datetime
+
+
+def read_inpur():
```
Comment (abuccts) [inline contrib/remote-dev/start.py]: `read_inpur` -> `read_input`

File: contrib/remote-dev/start.py [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+  f1 = open("conf/" + conf_file + ".template", 'r+')
+  f2 = open(userdir + conf_file + ".yaml", 'w+')
+  for line in f1.readlines():
+    f2.write(replace(line))
+  f1.close()
+  f2.close()
+
+# display PAI cluster
+print_output("Display PAI cluster")
+subprocess.call("opai cluster list", shell=is_shell)
```
Comment (abuccts) [inline contrib/remote-dev/start.py]: where did you install `opai`?

File: contrib/remote-dev/start.py [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+
+def replace(line):
+  tt = re.sub("\$username", username, line)
+  tt = re.sub("\$password", password, tt)
+  tt = re.sub("\$serverip", serverip, tt)
+  return tt
+
+
+# read env vars
+print_output("Read .env")
```
Comment (abuccts) [inline contrib/remote-dev/start.py]: please add a main function

File: contrib/remote-dev/start.py [hunk 20]
```diff
@@ -0,0 +1,124 @@
+import re
```
Comment (abuccts) [inline contrib/remote-dev/start.py]: license header

File: contrib/remote-dev/README.md [hunk 21]
```diff
@@ -0,0 +1,183 @@
...
+
+Users should edit the resources in `resourcePerInstance`, and can also change the docker image in `resourcePerInstance`. Please make sure your customized docker image has openssh-server. Users can also change `commands` if they want to mount storage or install dependencies in the container.
+
+We suppose that your training data is stored in a storage server and managed by PAI's [Team Wise Storage](https://github.com/microsoft/pai/tree/master/contrib/storage_plugin). You should change `commands` to mount your data server.
+
+You can use the following commands to mount external storage.
+
+```sh
+# mount NFS
+mount -t <nfsversion> <hostip>:<sharepath> <mountpoint>
```
Comment (abuccts) [inline contrib/remote-dev/README.md]: `<nfsversion> -> [nfs|nfs4]`

File: contrib/remote-dev/README.md [hunk 22]
```diff
@@ -0,0 +1,183 @@
...
+│   ├── clusters.template        # PAI cluster template conf
+│   └── job.template             # PAI template job
+└── start.py                     # start script
+```
+
+## 3. Dependencies
+
+- This subproject depends on [PAI's Python SDK](https://github.com/microsoft/pai/tree/master/contrib/python-sdk), please install and test the SDK first.
+
+- The `start.py` script requires python3, and we only tested it on `py3.5+` environment.
```
Comment (abuccts) [inline contrib/remote-dev/README.md]: could you make it compatible with both py2 and py3?

File: contrib/remote-dev/conf/job.template (hunk 11 above)
Comment (TongboZhang) [reply contrib/remote-dev/conf/job.template]: Can this work now? I remember that this cannot work before.

File: contrib/remote-dev/README.md (hunk 22 above)
Comment (TongboZhang) [reply contrib/remote-dev/README.md]: Nope, because python-sdk only support py3.

File: contrib/remote-dev/conf/job.template (hunk 11 above)
Comment (abuccts) [reply contrib/remote-dev/conf/job.template]: this is comment in yaml, not shell command
the comment is read by user, not shell

File: contrib/remote-dev/README.md (hunk 22 above)
Comment (abuccts) [reply contrib/remote-dev/README.md]: ...

**Comment by TongboZhang** [comment]
I have refactored the code, please check it out. Thanks a lot!
@abuccts  @Binyang2014

File: contrib/remote-dev-tool/README.md [hunk 23]
```diff
@@ -0,0 +1,88 @@
+# Remote-dev-tool
+
+## 1. Overview
+
+**Remote-dev-tool** allows users to share a local folder and mount it in PAI's container.
+And also support submit a job from local YAML.
```
Comment (abuccts) [inline contrib/remote-dev-tool/README.md]: "also support" -> "it also supports to"

File: contrib/remote-dev-tool/conf/exports.template [hunk 24]
```diff
@@ -0,0 +1 @@
+/workspace *(rw,fsid=0,sync,no_root_squash,no_subtree_check,anonuid=1000,anongid=1000)
```
Comment (abuccts) [inline contrib/remote-dev-tool/conf/exports.template]: is this file needed?

File: contrib/remote-dev-tool/remote-dev-tool.py [hunk 25]
```diff
@@ -0,0 +1,274 @@
...
+    parser.add_argument("-g", "--getssh", metavar="job_name", help="get ssh info", dest="job_name")
+    parser.add_argument("-s", "--submit", metavar='job_path', help="submit local job", dest="job_path")
+    parser.add_argument("-sh", "--share", metavar="share_path", help="share local folder", dest="share_path")
+    parser.add_argument("-v", "--verbose", action="store_true", help="verbose mode", dest='need_verbose')
+    args = parser.parse_args()
+    if len(sys.argv) < 2 or (len(sys.argv) == 2 and args.need_verbose is True):
+        parser.print_help()
+        parser.exit(-1)
+    if args.job_name is not None and args.job_path is not None:
+        LOGGER.info("cannot use --submit and --getssh at the same time")
```
Comment (abuccts) [inline contrib/remote-dev-tool/remote-dev-tool.py]: it's `error` instead of `info`

File: contrib/remote-dev-tool/remote-dev-tool.py [hunk 26]
```diff
@@ -0,0 +1,274 @@
...
+        parser.print_help()
+        parser.exit(-1)
+    if args.job_name is None and args.job_path is None:
+        LOGGER.info("cannot use without both --submit and --getssh")
```
Comment (abuccts) [inline contrib/remote-dev-tool/remote-dev-tool.py]: same

File: contrib/remote-dev-tool/remote-dev-tool.py [hunk 27]
```diff
@@ -0,0 +1,274 @@
...
+        parser.exit(-1)
+
+    return args
+
+
+def get_ssh_info(job_name):
+    global LOGGER, SSH_INFO, USER_DIR, JOB_NAME
+    LOGGER.debug("get SSH info")
+    JOB_NAME = job_name
+    pai_dir = USER_DIR + ".openpai/"
```
Comment (abuccts) [inline contrib/remote-dev-tool/remote-dev-tool.py]: use `os.path.join`

File: contrib/remote-dev-tool/README.md [hunk 28]
```diff
@@ -0,0 +1,88 @@
...
+Remote Development Tool
+
+optional arguments:
+  -h, --help                        
+                                    show this help message and exit
+  -g job_name, --getssh job_name
+                                    get ssh info
+  -s path, --submit path
+                                    submit local job
+  -sh path, --share path
```
Comment (abuccts) [inline contrib/remote-dev-tool/README.md]: maybe you can change `-s` to `-c` for config file and `-sh` to `-s`

File: contrib/remote-dev-tool/conf/exports.template (hunk 24 above)
Comment (TongboZhang) [reply contrib/remote-dev-tool/conf/exports.template]: This file is needed for the NFS docker in ubuntu.

File: contrib/remote-dev-tool/remote-dev-tool.py (hunk 25 above)
Comment (TongboZhang) [reply contrib/remote-dev-tool/remote-dev-tool.py]: Changed

File: contrib/remote-dev-tool/remote-dev-tool.py (hunk 26 above)
Comment (TongboZhang) [reply contrib/remote-dev-tool/remote-dev-tool.py]: Changed

File: contrib/remote-dev-tool/remote-dev-tool.py (hunk 27 above)
Comment (TongboZhang) [reply contrib/remote-dev-tool/remote-dev-tool.py]: Changed
